### ISSUE TITLE
Clarifying the allowed roles on an li element

### ISSUE THREAD

**Issue by scottaohara** [validator/validator#1368]
[ARIA in HTML](https://github.com/w3c/html-aria/pull/410) is clarifying the allowed roles on `li` elements.
><a>No `role`</a></strong> if the parent list element has an implicit  or explicit `list` role.
>
 >Otherwise, <a><strong>any `role`</strong></a>.
> NOTE: See <a >`ul`</a>, <a >`ol`</a>, or <a>`menu`</a> for allowed roles for list elements.

This update **disallows any role** on an `li` element so long as it is a child of a `ul`, `ol` or `menu` element and that element is exposed as an implicit or explicit role=list. Specifying a role of `listitem` on an `li` would produce a **warning** for unnecessary use of a redundant role.

However, if the parent `ul`, `ol` or `menu` has had its implicit list role overwritten, then the `li` may have **any role** so long as that role is an allowed role of its ancestor elements.

For example, the following `li` may have a `role=button` specified as the parent list element as a `role=none`
```html
<ul role=none>
  <li role="button" ...>...</li>
</ul>
```

However, as the following `ul` has a `role=menu` specified, the `role=button` is not allowed.  The only allowed roles for the `li` would be `menuitem`, `menuitemcheckbox` or `menuitemradio`.
```html
<ul role=menu>
  <li role=button>...</button>
</ul>
```

similar to the following, as while the `ul` has a `role=none`, it is a child of a `role=tablist`, so therefor the `li` may only have a role that is an expected descendant of a `tablist`.

```html
<div role=tablist>
  <ul role=none>
    <li role=button>...</li>
  </ul>
</div>
```

Finally, even though the following would be flagged for invalid HTML, the `li` is again allowed any role as it is exposed as a `generic` role when not correctly nested within a list element:

```html
<article>
   ....
   <li role=paragraph>...</li>
   ...
</article>
```

Additional [test cases provided here](https://w3c.github.io/html-aria/tests/li-element-roles.html).

**Comment by sideshowbarker** [comment]
> However, as the following `ul` has a `role=menu` specified, the `role=button` is not allowed. The only allowed roles for the `li` would be `menuitem`, `menuitemcheckbox` or `menuitemradio`.
> 
>```html
> <ul role=menu>
>   <li role=button>...</button>
> </ul>
>```

Can you point me to exactly where in the ARIA in HTML that requirement is stated?

At https://w3c.github.io/html-aria/#el-li the complete relevant requirements are in full:

> [No role](https://w3c.github.io/html-aria/#dfn-no-role) other than [`listitem`](https://w3c.github.io/html-aria/#index-aria-listitem), which is NOT RECOMMENDED, if the parent `list` element has an implicit or explicit `list` role.
> 
> Otherwise, [any role](https://w3c.github.io/html-aria/#dfn-any-role) if the parent `list` element does not expose an implicit or explicit `list` role.

Based just on those cited requirements, I would expect that `li` is allowed to have any role — per the _“the parent `list` element does not expose an implicit or explicit `list` role”_ condition.

What am I missing?

> similar to the following, as while the `ul` has a `role=none`, it is a child of a `role=tablist`, so therefor the `li` may only have a role that is an expected descendant of a `tablist`.
> 
>```html
> <div role=tablist>
>   <ul role=none>
>     <li role=button>...</li>
>   </ul>
> </div>
>```

For that also, could you point me to exactly where in the ARIA in HTML that requirement is stated?

…because based just on the `li` requirements above cited from the spec, I’d expect that `li` is allowed to have any role.

**Comment by pkra** [comment]
> Can you point me to exactly where in the ARIA in HTML that requirement is stated?

It follows from ARIA: `button` is not an Allowed Accessibility Child Role for `menu`, cf. https://w3c.github.io/aria/#menu

**Comment by sideshowbarker** [comment]
> > Can you point me to exactly where in the ARIA in HTML that requirement is stated?
> 
> It follows from ARIA: `button` is not an Allowed Accessibility Child Role for `menu`, cf. [w3c.github.io/aria#menu](https://w3c.github.io/aria/#menu)

OK, thanks — but I think those additional requirements should rightly also be explicitly stated (or referenced) in the ARIA in HTML spec too — at point of use at https://w3c.github.io/html-aria/#el-li.

I think the omission of an explicit such statement or reference at point of use in the ARIA in HTML spec would (mis)lead other checker implementors (not just me) into reasonably assuming that the requirements stated at that point in the ARIA in HTML spec are a complete statement of all the requirements.

**Comment by pkra** [comment]
@sideshowbarker I think this is primarily covered in Section 5 https://w3c.github.io/html-aria/#allowed-descendants-of-aria-roles 

However aria-in-html does not seem to use the current ARIA language (after https://github.com/w3c/aria/pull/1454). I filed https://github.com/w3c/html-aria/issues/561 to try to match this in aria-in-html.

**Comment by sideshowbarker** [comment]
> [@sideshowbarker](https://github.com/sideshowbarker) I think this is primarily covered in Section 5 [w3c.github.io/html-aria#allowed-descendants-of-aria-roles](https://w3c.github.io/html-aria/#allowed-descendants-of-aria-roles)

Ah indeed — thanks much, that looks like what I need.

> However aria-in-html does not seem to use the current ARIA language (after [w3c/aria#1454](https://github.com/w3c/aria/pull/1454)). I filed [w3c/html-aria#561](https://github.com/w3c/html-aria/issues/561) to try to match this in aria-in-html.

Subscribed.
